### PR TITLE
Fixed #30828 -- Added add_relations and remove_relations.

### PR DESCRIPTION
**Pull request by davidfstr**
Trac ticket: https://code.djangoproject.com/ticket/30828

### CODE REVIEW

**Comment by davidfstr** [comment]
Above it says the flake8 test failed, but running `flake8` on my local machine produces no bad output. The above failure provides no detail in its output, so I don't see a good way to investigate further.

Presumably if someone else can get a clean run of `flake8` on their machine when checking out the tip of this branch then the flake8 test failure can be ignored as bogus.

**Comment by charettes** [comment]
[Here's the output](https://djangoci.com/job/flake8/2203/console).

```
+ flake8
./django/db/models/fields/related_descriptors.py:817:47: F821 undefined name 'Set'
./django/db/models/fields/related_descriptors.py:817:47: F821 undefined name 'Tuple'
./django/db/models/fields/related_descriptors.py:817:47: F821 undefined name 'PK'
./django/db/models/fields/related_descriptors.py:817:47: F821 undefined name 'PK'
```

File: django/db/models/fields/related_descriptors.py [hunk 1]
```diff
@@ -798,6 +798,81 @@ def _get_set_deprecation_msg_params(self):
             self.rel.get_accessor_name() if self.reverse else self.field.name,
         )
 
+    def add_relations(self, relations, *, assert_no_collisions=False):
+        """
+        Creates many (M1, M2) relations with O(1) database queries.
+
+        Any requested relations that already exist will be left alone.
```
Comment (orf) [inline django/db/models/fields/related_descriptors.py]: "Any requested relations that already exist will not be altered." is perhaps better wording?

File: django/db/models/fields/related_descriptors.py [hunk 2]
```diff
@@ -798,6 +798,81 @@ def _get_set_deprecation_msg_params(self):
...
+
+        If you assert that none of the requested relations already exist,
+        you can pass assert_no_collisions=True to save 1 database query.
+        """
+        if len(relations) == 0:
+            return
+
+        (M1_M2, m1_pk_field_name, m2_pk_field_name, m1_m2_pk_tuples) = \
```
Comment (orf) [inline django/db/models/fields/related_descriptors.py]: The parenthesis here are unneeded

File: django/db/models/fields/related_descriptors.py [hunk 3]
```diff
@@ -798,6 +798,81 @@ def _get_set_deprecation_msg_params(self):
...
+            existing_m1_m2_pk_tuples = {
+                (getattr(m1_m2, m1_pk_field_name), getattr(m1_m2, m2_pk_field_name))
+                for m1_m2 in
+                M1_M2.objects.filter(q)
+            }
+
+        M1_M2.objects.bulk_create([
+            M1_M2(**{m1_pk_field_name: m1_pk, m2_pk_field_name: m2_pk})
+            for (m1_pk, m2_pk) in m1_m2_pk_tuples
+            if (m1_pk, m2_pk) not in existing_m1_m2_pk_tuples
```
Comment (orf) [inline django/db/models/fields/related_descriptors.py]: Could we leverage the `ignore_conflicts` parameter to not require the collisions query on databases that support it?

File: tests/many_to_many/tests.py [hunk 4]
```diff
@@ -165,6 +166,116 @@ def test_slow_add_ignore_conflicts(self):
                 self.a1.publications.add(self.p1)
         mocked.assert_called_once()
 
+    def test_add_relations(self):
+        self.assertIn(self.p1, self.a1.publications.all())
```
Comment (orf) [inline tests/many_to_many/tests.py]: `self.assertQuerysetEqual(self.a1.publications.all(), [self.p1])` is a way to remove these repetitive assertions in all the tests below

File: tests/many_to_many/tests.py [hunk 5]
```diff
@@ -165,6 +166,116 @@ def test_slow_add_ignore_conflicts(self):
...
+            Article.publications.add_relations([
+                (self.a3, self.p3),
+            ], assert_no_collisions=True)
+        self.assertIn(self.p3, self.a3.publications.all())
+
+        # Collision! Should raise IntegrityError. Should affect no rows.
+        self.assertNotIn(self.p4, self.a4.publications.all())
+        self.assertIn(self.p1, self.a1.publications.all())
+        with transaction.atomic():
+            with self.assertRaisesMessage(IntegrityError, ''):
```
Comment (orf) [inline tests/many_to_many/tests.py]: I think this should be in a separate test, if possible?

File: django/db/models/fields/related_descriptors.py (hunk 2 above)
Comment (orf) [reply django/db/models/fields/related_descriptors.py]: Can we also rename `M1_M2` to `through_model` or somesuch?

File: docs/ref/models/relations.txt [hunk 6]
```diff
@@ -205,3 +205,88 @@ Related objects reference
...
+              # ...
+              pass
+
+          class Pizza(models.Model):
+              toppings = models.ManyToManyField(Topping)
+
+    In this example, the methods below will be available both on
+    ``Topping.pizza_set`` and on ``Pizza.toppings``.
+
+    .. method:: add_relations(relations, *, assert_no_collisions=False)
```
Comment (orf) [inline docs/ref/models/relations.txt]: I think it should be noted somewhere here that this won't work with custom through models, unless the fields have a default value set for all non-pk columns?

File: django/db/models/fields/related_descriptors.py [hunk 7]
```diff
@@ -798,6 +798,81 @@ def _get_set_deprecation_msg_params(self):
...
+        ])
+
+    def remove_relations(self, relations):
+        """
+        Deletes many (M1, M2) relations with O(1) database queries.
+
+        Any requested relations that did not exist before will be skipped.
+        """
+        if len(relations) == 0:
```
Comment (orf) [inline django/db/models/fields/related_descriptors.py]: `if not relations:` is a sufficient check here I think

**Comment by felixxm** [comment]
@davidfstr Thanks for this patch :+1: , however I agree with Simon. I don't see much benefits over calling `bulk_create()`. IMO we should change this ticket to a documentation issue and add a paragraph about `ManyToManyField`'s with a reverse relationship to the _"Insert in bulk"_ section (`docs/topics/db/optimization.txt`).

File: django/db/models/fields/related_descriptors.py (hunk 1 above)
Comment (davidfstr) [reply django/db/models/fields/related_descriptors.py]: Will do.

File: django/db/models/fields/related_descriptors.py (hunk 2 above)
Comment (davidfstr) [reply django/db/models/fields/related_descriptors.py]: I like providing parens when working with tuples. Especially long ones like this. Does Django have a local convention that avoids parens for tuples in general?

Sure I can make use `through_model` instead.

File: django/db/models/fields/related_descriptors.py (hunk 3 above)
Comment (davidfstr) [reply django/db/models/fields/related_descriptors.py]: I didn't know about that parameter. I'll do some reading on it...

File: django/db/models/fields/related_descriptors.py (hunk 7 above)
Comment (davidfstr) [reply django/db/models/fields/related_descriptors.py]: I usually only use truthiness checks on booleans.

File: docs/ref/models/relations.txt (hunk 6 above)
Comment (davidfstr) [reply docs/ref/models/relations.txt]: Agreed. Same also holds for methods like `m1.add(*[a, b, c])` I believe. I'll check their docs.

File: tests/many_to_many/tests.py (hunk 5 above)
Comment (davidfstr) [reply tests/many_to_many/tests.py]: Will split test to:

* test_add_relations_with_assert_no_collisions_and_no_collision
* test_add_relations_with_assert_no_collisions_and_collision

**Comment by davidfstr** [comment]
I've created a separate PR https://github.com/django/django/pull/11948 that uses a documentation workaround to illustrate how to bulk associate/disassociate lots of related object pairs.

Please comment on the documentation change there.

**Comment by carltongibson** [comment]
Hi @davidfstr. Thanks for this. Given Simon and Mariusz' assessment, I'm going to close this so we can focus on #11948. 

For me: 

* The few lines _boilerplate™_ is not a problem. Scripting them up in `bulk_create()` context is easy enough, but more, working with through models in this way is an important facet of the ORM. I'm not sure hiding it here helps in the long run. 
* Then, I also think the similarity between `add()` and `add_relation()` (&co) would be a point of confusion. I think we're better off served keeping the base API as it is, and then having folks [work _through_ the appropriate docs](https://docs.djangoproject.com/en/2.2/topics/db/models/#intermediary-manytomany) when they need them. 

I hope that makes sense. I think we're already all agreed on the docs approach anyway...

**Pull request closed without merging**